lib/src/signed_video_h26x_auth.c

```diff
@@ -353,6 +459,79 @@ set_validation_status_of_items_used_in_gop_hash(h26x_nalu_list_t *nalu_list, cha
   return num_marked_items;
 }
 
+/**
+ * Verifies the integrity of the GOP hash in the video, ensuring that the data
+ * within the GOP is authentic and complete. Updates the expected and received
+ * NALU counts, and returns true if the verification is successful.
+ */
+static bool
+verify_hashes_with_sei(signed_video_t *self, int *num_expected_nalus, int *num_received_nalus)
+{
+  assert(self);
+
+  int num_expected_hashes = -1;
+  int num_received_hashes = -1;
+  char validation_status = 'P';
+  h26x_nalu_list_t *nalu_list = self->nalu_list;
+  h26x_nalu_list_item_t *sei = h26x_nalu_list_get_next_sei_item(nalu_list);
+  svrc_t status = SV_UNKNOWN_FAILURE;
+
+  // Mark all the NALUs that belong to the GOP to verify.
+  add_hashes_for_computing_gop_hash(self, sei);
+
+  // The verified_signature_hash indicates if the signature is verified.
+  // If the signature hash is verified, the GOP hash can be verified as well.
+  // If the signature hash is not verified, it means the SEI is corrupted, and the whole GOP status
+  // is determined by the verified_signature_hash.
+  switch (self->gop_info->verified_signature_hash) {
+    case 1:
+      // The signature is verified, GOP hash can be verified.
+      if (!verify_gop_hash(self)) {
+        // If the signature is verified but GOP hash is not, continue validation with the
+        // hash list if it is accessible.
+        if (self->gop_info->list_idx > 0) {
+          return verify_hashes_with_hash_list(self, num_expected_nalus, num_received_nalus);
+        }
+      }
+      validation_status = self->gop_info->successfully_verified_gop_hash ? '.' : 'N';
+      break;
+    case 0:
+      validation_status = 'N';
+      break;
+    case -1:
+    default:
+      // An error occurred when verifying the GOP hash. Verify without a SEI.
+      validation_status = 'E';
+      return verify_hashes_without_sei(self);
+  }
+
+  // The number of hashes part of the GOP hash was transmitted in the SEI.
+  num_expected_hashes = (int)self->gop_info->num_sent_nalus;
```

Comment:
If ´self->gop_info->verified_signature_hash = 0´ you cannot trust the content in the SEI, so you do not know how many NALUs that were sent.
Basically, all operations below only holds if the SEI can be verified successfully.